doc/user-guide/testing.rst

```diff
@@ -0,0 +1,125 @@
+.. _testing:
+
+Testing your code
+=================
+
+.. ipython:: python
+    :suppress:
+
+    import numpy as np
+    import pandas as pd
+    import xarray as xr
+
+    np.random.seed(123456)
+
+.. _asserts:
+
+Asserts
+-------
+
+TODO
+
+.. _hypothesis:
+
+Hypothesis testing
+------------------
+
+.. note::
+
+  Testing with hypothesis is a fairly advanced topic. Before reading this section it is recommended that you take a look
+  at our guide to xarray's data structures, are familiar with conventional unit testing in pytest, and have seen the
+  hypothesis library documentation.
+
+``Hypothesis`` is a powerful library for property-based testing.
+Instead of writing tests for one example at a time, it allows you to write tests parameterized by a source of many
+dynamically generated examples. For example you might have written a test which you wish to be parameterized by the set
+of all possible ``integers()``.
+
+Property-based testing is extremely powerful, because (unlike more conventional example-based testing) it can find bugs
+that you did not even think to look for!
+
+Strategies
+~~~~~~~~~~
+
+Each source of examples is called a "strategy", and xarray provides a range of custom strategies which produce xarray
+data structures containing arbitrary data. You can use these to efficiently test downstream code,
+quickly ensuring that your code can handle xarray objects of all possible structures and contents.
+
+These strategies are accessible in the :py:module::`xarray.testing.strategies` module, which provides
+
+.. currentmodule:: xarray
+
+.. autosummary::
+
+   testing.strategies.valid_dtypes
+   testing.strategies.np_arrays
```

Comment:
The purpose of that is because xarray only accepts certain dtypes right? I don't have strong feelings about this though, except that users should have all the tools they need to build their own valid xarray strategies.

Reply:
I think `xarray` would be fine with almost every dtype (except maybe the structured dtypes), but `sparse` in particular is very restricted.